packages/teleport/src/Discover/Kubernetes/TestConnection/TestConnection.tsx

```diff
@@ -0,0 +1,323 @@
+import {
+  Header,
+  ActionButtons,
+  TextIcon,
+  HeaderSubtitle,
+  Mark,
+  ReadOnlyYamlEditor,
+} from '../../Shared';
+import { ruleConnectionDiagnostic } from '../../templates';
+
+import { useTestConnection, State } from './useTestConnection';
+
+import type { AgentStepProps } from '../../types';
+
+export default function Container(props: AgentStepProps) {
+  const ctx = useTeleport();
+  const state = useTestConnection({ ctx, props });
+
+  return <TestConnection {...state} />;
+}
+
+export function TestConnection({
+  attempt,
+  runConnectionDiagnostic,
+  diagnosis,
+  nextStep,
+  canTestConnection,
+  kube,
+  authType,
+  username,
+  clusterId,
+}: State) {
+  const userOpts = kube.users.map(l => ({ value: l, label: l }));
+  const groupOpts = kube.groups.map(l => ({ value: l, label: l }));
+
+  const errorRefs = useRef<HTMLDivElement[]>([]);
+
+  const [namespace, setNamespace] = useState('default');
+  const [selectedGroups, setSelectedGroups] = useState(groupOpts);
+
+  // Always default it to either teleport username or from one of users defined
+  // from previous step.
+  const [selectedUser, setSelectedUser] = useState(() =>
+    userOpts.length > 0 ? userOpts[0] : { value: username, label: username }
+  );
+
+  const { hostname, port } = window.document.location;
+  const host = `${hostname}:${port || '443'}`;
+  const authSpec =
+    authType === 'local' ? `--auth=${authType} --user=${username} ` : '';
+  const tshLoginCmd = `tsh login --proxy=${host} ${authSpec}${clusterId}`;
+
+  let $diagnosisStateComponent;
+  if (attempt.status === 'processing') {
+    $diagnosisStateComponent = (
+      <TextIcon>
+        <Icons.Restore fontSize={4} />
+        Testing in-progress
+      </TextIcon>
+    );
+  } else if (attempt.status === 'failed' || (diagnosis && !diagnosis.success)) {
+    $diagnosisStateComponent = (
+      <TextIcon>
+        <Icons.Warning ml={1} color="danger" />
+        Testing failed
+      </TextIcon>
+    );
+  } else if (attempt.status === 'success' && diagnosis?.success) {
+    $diagnosisStateComponent = (
+      <TextIcon>
+        <Icons.CircleCheck ml={1} color="success" />
+        Testing complete
+      </TextIcon>
+    );
+  }
+
+  const showDiagnosisOutput = !!diagnosis || attempt.status === 'failed';
+
+  function handleTestConnection(validator: Validator) {
+    if (!validator.validate()) {
+      return;
+    }
+
+    runConnectionDiagnostic({
+      namespace,
+      user: selectedUser.value,
+      groups: selectedGroups.map(g => g.value),
+    });
+  }
+
+  const errBtnTxt = 'Click for extra details';
+  function toggleTraceErrorStates(index: number) {
+    const el = errorRefs.current[index];
+
+    const textEl = el.lastElementChild as HTMLElement;
+    textEl.style.display = textEl.style.display === 'none' ? 'block' : 'none';
+
+    const btnEl = el.firstElementChild as HTMLElement;
+    btnEl.textContent =
+      btnEl.textContent === errBtnTxt ? 'Hide details' : errBtnTxt;
+  }
```

Comment:
I think we shouldn't mutate DOM directly (its React job). Instead of that we can extract this logic to separate component with local state:
```suggestion
type ErrorWithDetailsProps = {
  error: string;
  details: string;
};

const ErrorWithDetails = ({ details, error }: ErrorWithDetailsProps) => {
  const [toggle, setToggle] = useState(false);

  return (
    <TextIcon css={{ alignItems: 'baseline' }}>
      <Icons.CircleCross mr={1} color="danger" />
      <div>
        <div>{details}</div>
        <div>
          <ButtonShowMore onClick={() => setToggle(p => !p)}>
            {toggle ? 'Click for extra details' : 'Hide details'} details
          </ButtonShowMore>
          {toggle && <div>{error}</div>}
        </div>
      </div>
    </TextIcon>
  );
};
```